url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/FCSAgent.java

```diff
@@ -0,0 +1,373 @@
+		PENDING, STARTED, LOADED
+	};
+
+	public FCSAgent(String name) {
+		super();
+		this.name = name;
+		this.nests = new ArrayList<Nest>();
+		this.orders = new ArrayList<Order>();
+		binsSet = false;
+		binsToAdd = new ArrayList<PartType>();
+		state = myState.PENDING;
+	}
+
+	public FCSAgent() {
+		super();
+		this.name = "FCS Agent";
+		binsSet = false;
+		binsToAdd = new ArrayList<PartType>();
+		state = myState.PENDING;
+	}
+
+	@Override
+	public void msgAddKitsToQueue(Order o) {
+		print("Received new order");
+		orders.add(o);
+		if (fcs != null) {
+			fcs.updateQueue((ArrayList<Order>) orders);
+		}
+		if (state == myState.PENDING) {
+			state = myState.STARTED;
+		}
+		stateChanged();
+	}
+
+	public void msgSetPartsRobotDropChance(Float c) {
+		partsRobot.msgSetDropChance(c);
+		stateChanged();
+	}
+
+	@Override
+	public void msgStopMakingKit(Order o) {
+		print("Received msgStopMakingKit");
+		synchronized (orders) {
+			for (Order order : orders) {
+				if (order.equals(o)) {
+					o.cancel = true;
+					if (fcs != null) {
+						fcs.updateQueue((ArrayList<Order>) orders);
+					}
+				}
+			}
+		}
+
+		// resetCell(o);
+		stateChanged();
+	}
+
+	@Override
+	public void msgStartProduction() {
+		print("Received msgStartProduction");
+		state = myState.STARTED;
+		stateChanged();
+	}
+
+	@Override
+	public void msgAddNewPartType(PartType part) {
+		print("Received msgAddNewPartType");
+		binsToAdd.add(part);
+		stateChanged();
+	}
+
+	@Override
+	public void msgShippedKit() {
+		fcs.shippedKit();
+	}
+
+	@Override
+	public void msgOrderFinished() {
+		print("Received msgOrderFinished");
+		numOrdersFinished++;
+		System.out.println("Order " + numOrdersFinished + " Done!!!!");
+		synchronized (orders) {
+			for (Order o : orders) {
+				if (o.state == Order.orderState.ORDERED) {
+					orders.remove(o);
+					if (fcs != null) {
+						fcs.updateQueue((ArrayList<Order>) orders);
+					}
+					// resetCell(o);
+					break;
+				}
+			}
+		}
+		state = myState.STARTED;
+		stateChanged();
+	}
+
+	@Override
+	public void msgBreakLane(int laneNumber) {
+		((NestAgent) nests.get(laneNumber)).lane.msgBreakThis();
+	}
+
+	@Override
+	public boolean pickAndExecuteAnAction() {
+		// print("I'm scheduling stuff");
+		if (state == myState.STARTED) {
+			if (orders.size() == 0) {
+				resetCell();
+			}
+			if (!binsSet && gantry != null) {
+				initializeBins();
+				return true;
+			}
+			if (binsToAdd.size() > partTypeCount && gantry != null) {
+				addBin();
+				return true;
+			}
+			if (!orders.isEmpty()) {
+				print("Checking orders");
+				synchronized (orders) {
+					for (Order o : orders) {
+						if (o.cancel) {
+							cancelOrder(o);
+							return true;
+						}
+					}
+					for (Order o : orders) {
+						if (o.state == Order.orderState.PENDING) {
+							placeOrder(o);
+							return true;
+						}
+					}
+				}
+				print("Found no orders to process");
+			}
+		}
+		return false;
+	}
+
+	public void placeOrder(Order o) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 18. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 205-220, lines 221-243 into separate methods.